### PR TITLE
New Boiler Ability: Emit Acid

### PR DESCRIPTION
**Pull request by Surrealaser**
:cl: by Surrealistik
add: Boilers can now emit a cloud of acid gas after a 0.5 second delay centred on their position. This costs 200 plasma, lasts 6 seconds, and has a cooldown of 60 seconds.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 1]
```diff
@@ -2500,3 +2500,71 @@
...
+	emitting_gas = TRUE //We gain bump movement immunity while we're emitting gas.
+	use_plasma(200)
+
+	if(!do_after(src, BOILER_GAS_CHANNEL_TIME, TRUE, 5, BUSY_ICON_HOSTILE))
+		smoke_system = new /datum/effect_system/smoke_spread/xeno_acid()
+		smoke_system.amount = 1
+		smoke_system.set_up(1, 0, get_turf(src))
+		to_chat(src, "<span class='xenodanger'>You abort emitting acid smoke, your expended plasma resulting in only a feeble wisp.</span>")
+		emitting_gas = FALSE
+		return
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: No cooldown after the failed attempt?

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 2]
```diff
@@ -2500,3 +2500,71 @@
...
+	dispense_gas()
+
+/mob/living/carbon/Xenomorph/Boiler/proc/boiler_gas_cooldown()
+	playsound(loc, 'sound/effects/xeno_newlarva.ogg', 25, 0)
+	to_chat(src, "<span class='xenodanger'>You feel your dorsal vents bristle with acid gas. You can use Emit Acid Smoke again.</span>")
+	update_action_button_icons()
+
+/mob/living/carbon/Xenomorph/Boiler/proc/dispense_gas(count = 1)
+	set waitfor = FALSE
+	while(count)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]:
```suggestion
	for(var/i = 1 to count)
```

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 3]
```diff
@@ -2500,3 +2500,71 @@
...
+			to_chat(src, "<span class='xenowarning'>You try to emit caustic gas but are disabled!</span>")
+			return
+		playsound(loc, 'sound/effects/smoke.ogg', 25)
+		var/turf/T = get_turf(src)
+		smoke_system = new /datum/effect_system/smoke_spread/xeno_acid()
+		smoke_system.amount = 1 + upgrade
+		smoke_system.set_up(4, 0, T, null, 6)
+		smoke_system.start()
+		T.visible_message("<span class='danger'>Caustic smoke billows from the glowing xenomorph!</span>")
+		count = max(0,count - 1)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]:
```suggestion
```

File: code/modules/mob/living/carbon/xenomorph/Powers.dm (hunk 1 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]: We'll see how it plays out

File: code/modules/mob/living/carbon/xenomorph/Castes/Boiler.dm [hunk 4]
```diff
@@ -280,3 +284,8 @@
 	bomb_cooldown = FALSE
 	to_chat(src, "<span class='notice'>You feel your toxin glands swell. You are able to bombard an area again.</span>")
 	update_action_button_icons()
+
+/mob/living/carbon/Xenomorph/Boiler/Bumped(atom/movable/AM as mob|obj)
+	if(emitting_gas) //We don't get bumped around
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Castes/Boiler.dm]: Do `anchored = TRUE` when emitting, and `anchored = FALSE` when not, instead, please.

Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Castes/Boiler.dm]: I really don't like using anchored as it can get wonky but sure

**Comment by DominikPanic** [comment]
#814 should be merged first

**Comment by Rohesie** [comment]
Tested. Works so far. Will be removed from TMing until 814 is in.

**Comment by DominikPanic** [comment]
@Rohesie Should remain DNM until #814 is in

**Comment by DominikPanic** [comment]
untested

```
runtime error: Undefined operation: 1 / the floor (15,90,1) (/turf/open/floor)
proc name: Initialize (/obj/effect/particle_effect/smoke/Initialize)
  source file: smoke.dm,34
  usr: Young Boiler (456) (/mob/living/carbon/Xenomorph/Boiler)
  src: the smoke (/obj/effect/particle_effect/smoke/xeno/burn)
  usr.loc: the floor (15,90,1) (/turf/open/floor)
  src.loc: null
  call stack:
the smoke (/obj/effect/particle_effect/smoke/xeno/burn): Initialize(0, 4, the floor (15,90,1) (/turf/open/floor), /datum/effect_system/smoke_spr... (/datum/effect_system/smoke_spread/xeno/acid))
Atoms (/datum/controller/subsystem/atoms): InitAtom(the smoke (/obj/effect/particle_effect/smoke/xeno/burn), /list (/list))
the smoke (/obj/effect/particle_effect/smoke/xeno/burn): New(0, 4, the floor (15,90,1) (/turf/open/floor), /datum/effect_system/smoke_spr... (/datum/effect_system/smoke_spread/xeno/acid))
the smoke (/obj/effect/particle_effect/smoke/xeno/burn): New(null, 4, the floor (15,90,1) (/turf/open/floor), /datum/effect_system/smoke_spr... (/datum/effect_system/smoke_spread/xeno/acid))
the smoke (/obj/effect/particle_effect/smoke/xeno/burn): New(null, 4, the floor (15,90,1) (/turf/open/floor), /datum/effect_system/smoke_spr... (/datum/effect_system/smoke_spread/xeno/acid))
/datum/effect_system/smoke_spr... (/datum/effect_system/smoke_spread/xeno/acid): start()
Young Boiler (456) (/mob/living/carbon/Xenomorph/Boiler): dispense gas(1)
Young Boiler (456) (/mob/living/carbon/Xenomorph/Boiler): emit acidgas()
Emit Acid (/datum/action/xeno_action/activable/emit_acidgas): use ability(the chair (/obj/structure/bed/chair))
Young Boiler (456) (/mob/living/carbon/Xenomorph/Boiler): MiddleClickOn(the chair (/obj/structure/bed/chair))
the chair (/obj/structure/bed/chair): Click(the floor (19,94,1) (/turf/open/floor), "mapwindow.map", "icon-x=13;icon-y=24;middle=1;s...")
```

File: code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm [hunk 5]
```diff
@@ -246,3 +247,86 @@
...
+	//give them fair warning
+	visible_message("<span class='danger'>Tufts of caustic smoke begin to billow from [src]!</span>", \
+	"<span class='xenodanger'>Caustic gas begins to billow from your mouth as you prepare to emit acid smoke. Keep still!</span>")
+
+	use_plasma(200)
+
+	anchored = TRUE //We gain bump movement immunity while we're emitting gas.
+	if(!do_after(src, BOILER_GAS_CHANNEL_TIME, TRUE, 5, BUSY_ICON_HOSTILE))
+		smoke_system = new /datum/effect_system/smoke_spread/xeno/acid
+		smoke_system.set_up(1, 0, get_turf(src))
```
Comment (Ghommie) [inline code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm]:
```suggestion
		smoke_system.set_up(1, get_turf(src))
```

File: code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm [hunk 6]
```diff
@@ -246,3 +247,86 @@
...
+		if(stagger) //If we got staggered, return
+			to_chat(src, "<span class='xenowarning'>You try to emit caustic gas but are staggered!</span>")
+			return
+		if(stunned || knocked_down)
+			to_chat(src, "<span class='xenowarning'>You try to emit caustic gas but are disabled!</span>")
+			return
+		playsound(loc, 'sound/effects/smoke.ogg', 25)
+		var/turf/T = get_turf(src)
+		smoke_system = new /datum/effect_system/smoke_spread/xeno/acid
+		smoke_system.set_up(4, 0, T, null, 6)
```
Comment (Ghommie) [inline code/modules/mob/living/carbon/xenomorph/castes/boiler/abilities_boiler.dm]:
```suggestion
		smoke_system.set_up(4, T, 6)
```

**Comment by Surrealaser** [comment]
@DominikPanic applied the recommended changes.

**Comment by DominikPanic** [comment]
> 
> 
> @DominikPanic applied the recommended changes.

Has this PR been tested locally?

**Comment by DominikPanic** [comment]
Stale

**Pull request closed without merging**